WeekFour.hs

```diff
@@ -0,0 +1,94 @@
+import Data.List 
+---------------------------------------------------------------------------------
+-- Exercise 1: Wholemeal programming
+---------------------------------------------------------------------------------
+fun1' :: [Integer] -> Integer
+fun1' = product . map (\x -> x-2) . filter (even)
+
+fun2' :: Integer -> Integer
+fun2' = sum . gatherNums
+
+gatherNums :: Integer -> [Integer]
+gatherNums 1    = [0]
+gatherNums n
+    | even n    = [n] ++ gatherNums (n `div` 2 )
+    | otherwise = gatherNums (3*n + 1)
+
+---------------------------------------------------------------------------------
+-- Exercise 2: Folding with trees 
+---------------------------------------------------------------------------------
+data Tree a = Leaf | Node Integer (Tree a) a (Tree a) deriving (Show, Eq)
+
+-- the main function to fold a list into a tree 
+foldTree :: [a] -> Tree a
+foldTree = foldr insertInTree Leaf  
+
+-- height function for inserting into a balanced binary tree, not ordered 
+height :: Tree a -> Integer
+height Leaf                  = 0
+height (Node _ Leaf _ Leaf)  = 0
+height (Node _ left _ right) = 1 + max (height left) (height right)
+
+insertInTree :: a -> Tree a -> Tree a 
+insertInTree x Leaf                  = Node 0 Leaf x Leaf
+insertInTree x (Node n left y right) = case compare (height left) (height right) of 
+    GT -> Node (n-1) left y (insertInTree x right)
+    EQ -> Node (h+1) left y (insertInTree x right)
+        where
+            h = height right 
+    LT -> Node (n-1) (insertInTree x left) y right
+
+-- Builds a balanced binary tree from a sorted list 
+buildFromSorted :: [a] -> Tree a 
+buildFromSorted []   = Leaf
+buildFromSorted list = Node (height) left (list!!middle) right 
+    where 
+        middle = (length list) `div` 2
+        left   = buildFromSorted $ take middle list 
+        right  = buildFromSorted $ drop (middle + 1) list 
+        height = floor $ logBase 2 (fromIntegral $ length list)
+
+---------------------------------------------------------------------------------
+-- Exercise 3: More folds
+---------------------------------------------------------------------------------
+-- The xor function below returns True if and only if there is an odd number of True
+-- values in the list supplied 
+xor :: [Bool] -> Bool
+xor [] = False
+xor (x:xs) = foldr (funcXOR) x xs 
+
+funcXOR :: Bool -> Bool -> Bool
+funcXOR a b = (a && (not b)) || ((not a) && b)
+
+-- The function below is equivalent to map, but using foldr instead 
+map' :: (a -> b) -> [a] -> [b]
+map' f [] = []
+map' f xs = foldr (\x xs -> (f x):xs) [] xs
```

Comment:
Again, you don't need case for empty list, you can write this function in single line.